In [2]:
import numpy as np
import pandas as pd

In [3]:
df = pd.read_csv("customers_messy_500.csv")
print(df.head())

  customer_id   age     city    income  score  purchases purchased
0       C1176  51.0  Colombo   60128.0  0.728        2.0         0
1       C1132  22.0    Galle   94989.0  0.387        4.0         0
2       C1006   NaN   Jaffna   81217.0  0.751        0.0         0
3       C1084  47.0  COLOMBO  103321.0  0.575        6.0         1
4       C1456   NaN    Galle   58237.0  0.798        8.0         1


In [4]:
print(df.tail())

    customer_id   age     city   income  score  purchases purchased
513       C1071   NaN    Kandy  47493.0  0.541        6.0         0
514       C1106  54.0    Galle  75680.0   0.67        0.0         0
515       C1270  31.0   JAFFNA  94152.0  0.509        8.0         0
516       C1435  23.0  Colombo  77433.0  0.338       10.0         0
517       C1102  55.0    Kandy  49434.0  0.616       10.0        NO


In [5]:
print(df.describe(include='all'))

       customer_id         age     city   income score   purchases purchased
count          517  481.000000      497      493   486  516.000000       516
unique         509         NaN       25      472   337         NaN        13
top          C1024         NaN  Colombo  18000.0  0.98         NaN         0
freq             2         NaN      170        4    14         NaN       304
mean           NaN   41.669439      NaN      NaN   NaN    5.418605       NaN
std            NaN   18.137923      NaN      NaN   NaN    3.474618       NaN
min            NaN   -3.000000      NaN      NaN   NaN    0.000000       NaN
25%            NaN   30.000000      NaN      NaN   NaN    2.000000       NaN
50%            NaN   41.000000      NaN      NaN   NaN    5.500000       NaN
75%            NaN   53.000000      NaN      NaN   NaN    8.000000       NaN
max            NaN  150.000000      NaN      NaN   NaN   11.000000       NaN


In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 518 entries, 0 to 517
Data columns (total 7 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   customer_id  517 non-null    object 
 1   age          481 non-null    float64
 2   city         497 non-null    object 
 3   income       493 non-null    object 
 4   score        486 non-null    object 
 5   purchases    516 non-null    float64
 6   purchased    516 non-null    object 
dtypes: float64(2), object(5)
memory usage: 28.5+ KB


In [7]:
df.shape

(518, 7)

In [8]:
df.dtypes

customer_id     object
age            float64
city            object
income          object
score           object
purchases      float64
purchased       object
dtype: object

In [9]:
print(df.isna().sum())

customer_id     1
age            37
city           21
income         25
score          32
purchases       2
purchased       2
dtype: int64


In [10]:
print(df.isnull().sum()/len(df)*100)

customer_id    0.193050
age            7.142857
city           4.054054
income         4.826255
score          6.177606
purchases      0.386100
purchased      0.386100
dtype: float64


In [11]:
empty_count = (df['city'].str.strip() == '').sum()
print(empty_count)


1


In [12]:
df.loc[0:9,["customer_id","age","city","score"]]

,customer_id,age,city,score
0,C1176,51.0,Colombo,0.728
1,C1132,22.0,Galle,0.387
2,C1006,NaN,Jaffna,0.751
3,C1084,47.0,COLOMBO,0.575
4,C1456,NaN,Galle,0.798
5,C1311,49.0,Colombo,0.676
6,C1333,57.0,Colombo,0.554
7,C1072,39.0,Colombo,0.549
8,C1184,41.0,Colombo,0.515
9,C1154,38.0,Negombo,0.616


In [13]:
df.iloc[0:10,0:4]

,customer_id,age,city,income
0,C1176,51.0,Colombo,60128.0
1,C1132,22.0,Galle,94989.0
2,C1006,NaN,Jaffna,81217.0
3,C1084,47.0,COLOMBO,103321.0
4,C1456,NaN,Galle,58237.0
5,C1311,49.0,Colombo,79461.0
6,C1333,57.0,Colombo,128793.0
7,C1072,39.0,Colombo,86247.0
8,C1184,41.0,Colombo,61311.0
9,C1154,38.0,Negombo,33377.0


In [14]:
data_series=df['score']
print(data_series)
print(type(data_series))

0      0.728
1      0.387
2      0.751
3      0.575
4      0.798
       ...  
513    0.541
514     0.67
515    0.509
516    0.338
517    0.616
Name: score, Length: 518, dtype: object
<class 'pandas.core.series.Series'>


In [15]:
data_frame=df[["age","score"]]
print(data_frame)

      age  score
0    51.0  0.728
1    22.0  0.387
2     NaN  0.751
3    47.0  0.575
4     NaN  0.798
..    ...    ...
513   NaN  0.541
514  54.0   0.67
515  31.0  0.509
516  23.0  0.338
517  55.0  0.616

[518 rows x 2 columns]


In [16]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 518 entries, 0 to 517
Data columns (total 7 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   customer_id  517 non-null    object 
 1   age          481 non-null    float64
 2   city         497 non-null    object 
 3   income       493 non-null    object 
 4   score        486 non-null    object 
 5   purchases    516 non-null    float64
 6   purchased    516 non-null    object 
dtypes: float64(2), object(5)
memory usage: 28.5+ KB


In [17]:
df["age"]=pd.to_numeric(df["age"],errors='coerce')

In [18]:
print(df["age"])

0      51.0
1      22.0
2       NaN
3      47.0
4       NaN
       ... 
513     NaN
514    54.0
515    31.0
516    23.0
517    55.0
Name: age, Length: 518, dtype: float64


In [19]:
df["score"]

0      0.728
1      0.387
2      0.751
3      0.575
4      0.798
       ...  
513    0.541
514     0.67
515    0.509
516    0.338
517    0.616
Name: score, Length: 518, dtype: object

In [20]:
df['score'] = (
    df['score']
    .astype(str)
    .str.replace('%', '', regex=False)
    .str.strip()
)

df['score'] = pd.to_numeric(df['score'], errors='coerce') / 100

print(df.head(50))

   customer_id    age       city     income    score  purchases purchased
0        C1176   51.0    Colombo    60128.0  0.00728        2.0         0
1        C1132   22.0      Galle    94989.0  0.00387        4.0         0
2        C1006    NaN     Jaffna    81217.0  0.00751        0.0         0
3        C1084   47.0    COLOMBO   103321.0  0.00575        6.0         1
4        C1456    NaN      Galle    58237.0  0.00798        8.0         1
5        C1311   49.0    Colombo    79461.0  0.00676        7.0         0
6        C1333   57.0   Colombo    128793.0  0.00554        5.0     false
7        C1072   39.0    Colombo    86247.0  0.00549        3.0         0
8        C1184   41.0   Colombo     61311.0  0.00515        4.0         0
9        C1154   38.0    Negombo    33377.0  0.00616        8.0         0
10       C1229   26.0      Kandy    18000.0  0.00513        3.0         0
11       C1304   43.0    Colombo    28625.0  0.00544       10.0     false
12       C1093   36.0      Kandy    63

In [21]:
print(df["score"].head(50))

0     0.00728
1     0.00387
2     0.00751
3     0.00575
4     0.00798
5     0.00676
6     0.00554
7     0.00549
8     0.00515
9     0.00616
10    0.00513
11    0.00544
12    0.00693
13    0.00519
14    0.00521
15    0.00671
16    0.00446
17    0.00794
18    0.00745
19    0.00529
20    0.47000
21    0.00731
22        NaN
23    0.00701
24    0.00712
25    0.00563
26    0.00501
27    0.56000
28    0.00590
29    0.00688
30    0.00663
31    0.68000
32    0.00691
33    0.00722
34    0.00237
35    0.00667
36    0.00338
37    0.00375
38    0.00412
39    0.00393
40    0.00769
41        NaN
42    0.00657
43    0.00552
44    0.00672
45    0.00671
46    0.00814
47        NaN
48    0.00643
49    0.00509
Name: score, dtype: float64


In [22]:
df['income'] = (
    df['income']
    .astype(str)
    .str.replace('LKR', '', regex=False)
    .str.replace(',', '', regex=False)
    .str.strip()
)
df['income'] = pd.to_numeric(df['income'], errors='coerce')



In [23]:
print(df[["age","score","income"]].dtypes)


age       float64
score     float64
income    float64
dtype: object


In [24]:
df["city"]=df["city"].str.strip()

In [25]:
df["city"]=df["city"].str.title()


In [26]:
print(df["city"].value_counts())

city
Colombo    201
Kandy      104
Galle       91
Jaffna      61
Negombo     35
Gale         2
Colmbo       2
             1
Name: count, dtype: int64


In [27]:
df["city"]=df["city"].replace({
    "Colmbo":"Colombo",
    "Gale":"Galle"
})

In [28]:
df['city'] = df['city'].replace('', np.nan)


In [29]:
print(df["city"].value_counts())

city
Colombo    203
Kandy      104
Galle       93
Jaffna      61
Negombo     35
Name: count, dtype: int64


In [30]:
df.isna().sum()

customer_id     1
age            37
city           22
income         25
score          32
purchases       2
purchased       2
dtype: int64

In [31]:
df.dropna()

,customer_id,age,city,income,score,purchases,purchased
0,C1176,51.0,Colombo,60128.0,0.00728,2.0,0
1,C1132,22.0,Galle,94989.0,0.00387,4.0,0
3,C1084,47.0,Colombo,103321.0,0.00575,6.0,1
5,C1311,49.0,Colombo,79461.0,0.00676,7.0,0
6,C1333,57.0,Colombo,128793.0,0.00554,5.0,false
...,...,...,...,...,...,...,...
509,C1466,25.0,Kandy,68044.0,0.00523,1.0,0
514,C1106,54.0,Galle,75680.0,0.00670,0.0,0
515,C1270,31.0,Jaffna,94152.0,0.00509,8.0,0
516,C1435,23.0,Colombo,77433.0,0.00338,10.0,0


In [32]:
df.isna().sum()

customer_id     1
age            37
city           22
income         25
score          32
purchases       2
purchased       2
dtype: int64

In [33]:
df["age"]=df["age"].fillna(df["age"].median())

In [34]:
df.isna().sum()

customer_id     1
age             0
city           22
income         25
score          32
purchases       2
purchased       2
dtype: int64

In [35]:
df["score"]=df["score"].fillna(df["score"].median())

In [36]:
df.isna().sum()

customer_id     1
age             0
city           22
income         25
score           0
purchases       2
purchased       2
dtype: int64

In [37]:
df["income"]=df["income"].fillna(df["income"].median())

In [38]:
df.isna().sum()

customer_id     1
age             0
city           22
income          0
score           0
purchases       2
purchased       2
dtype: int64

In [39]:
print(df['city'].isna().sum())          # how many missing
print(df['city'].isna().mean() * 100)   # % missing
print(df['city'].value_counts())

22
4.247104247104247
city
Colombo    203
Kandy      104
Galle       93
Jaffna      61
Negombo     35
Name: count, dtype: int64


In [40]:
df["city"]=df["city"].fillna(df["city"].mode()[0])

In [41]:
df.isna().sum()

customer_id    1
age            0
city           0
income         0
score          0
purchases      2
purchased      2
dtype: int64

In [42]:
df = df.dropna(subset=['customer_id'])
key_fields = ['age', 'score', 'income', 'city']
df = df.dropna(subset=key_fields, thresh=len(key_fields) - 2)  

In [43]:
print(df.isna().sum())

customer_id    0
age            0
city           0
income         0
score          0
purchases      1
purchased      1
dtype: int64


In [44]:
rows_before = df.shape[0]
print(rows_before)

517


In [45]:
df=df[(df["age"]>=18) & (df["age"]<=80)]

In [46]:
df = df[(df["score"]>=0) & (df["score"]<=1)]

In [47]:
df = df[(df["purchases"].notna()) & (df["purchases"] >= 0)]

In [48]:
row_after = df.shape[0]
print(row_after)

498


In [49]:
removed_row = rows_before - row_after
print(removed_row)

19


In [50]:
print(df.shape[0])

498


In [51]:
purchased_map = {
    'Yes': 1, 'YES': 1, 'yes': 1, 'true': 1, 'True': 1, 'TRUE': 1, 'y': 1, 'Y': 1, '1': 1, 1: 1,
    'No': 0, 'NO': 0, 'no': 0, 'false': 0, 'False': 0, 'FALSE': 0, 'n': 0, 'N': 0, '0': 0, 0: 0
}

df['purchased'] = df['purchased'].map(purchased_map)

In [52]:
df = df.dropna(subset=['purchased'])


In [53]:
df['purchased'] = df['purchased'].astype(int)

In [54]:
rows_before_dup = df.shape[0]
df = df.drop_duplicates()

In [55]:
dup_customer_ids = df['customer_id'].duplicated().sum()


In [56]:
df = df.drop_duplicates(subset=['customer_id'], keep='first')

In [57]:
print(f"Final number of rows: {df.shape[0]}")

Final number of rows: 490


In [58]:
df["score_pct"]=df["score"]*100
df.head(10)

,customer_id,age,city,income,score,purchases,purchased,score_pct
0,C1176,51.0,Colombo,60128.0,0.00728,2.0,0,0.728
1,C1132,22.0,Galle,94989.0,0.00387,4.0,0,0.387
2,C1006,41.0,Jaffna,81217.0,0.00751,0.0,0,0.751
3,C1084,47.0,Colombo,103321.0,0.00575,6.0,1,0.575
4,C1456,41.0,Galle,58237.0,0.00798,8.0,1,0.798
5,C1311,49.0,Colombo,79461.0,0.00676,7.0,0,0.676
6,C1333,57.0,Colombo,128793.0,0.00554,5.0,0,0.554
7,C1072,39.0,Colombo,86247.0,0.00549,3.0,0,0.549
8,C1184,41.0,Colombo,61311.0,0.00515,4.0,0,0.515
9,C1154,38.0,Negombo,33377.0,0.00616,8.0,0,0.616


In [59]:
print(df["income"].median())

75252.0


In [62]:
df["high_value"] = df["income"]>df["income"].median()
df.head(10)

,customer_id,age,city,income,score,purchases,purchased,score_pct,high_value
0,C1176,51.0,Colombo,60128.0,0.00728,2.0,0,0.728,False
1,C1132,22.0,Galle,94989.0,0.00387,4.0,0,0.387,True
2,C1006,41.0,Jaffna,81217.0,0.00751,0.0,0,0.751,True
3,C1084,47.0,Colombo,103321.0,0.00575,6.0,1,0.575,True
4,C1456,41.0,Galle,58237.0,0.00798,8.0,1,0.798,False
5,C1311,49.0,Colombo,79461.0,0.00676,7.0,0,0.676,True
6,C1333,57.0,Colombo,128793.0,0.00554,5.0,0,0.554,True
7,C1072,39.0,Colombo,86247.0,0.00549,3.0,0,0.549,True
8,C1184,41.0,Colombo,61311.0,0.00515,4.0,0,0.515,False
9,C1154,38.0,Negombo,33377.0,0.00616,8.0,0,0.616,False


In [66]:
print(df["score"].max())

0.93


In [67]:
df = df.sort_values("score",ascending=False)
df.head(10)

,customer_id,age,city,income,score,purchases,purchased,score_pct,high_value
260,C1038,31.0,Kandy,89072.0,0.93,9.0,1,93.0,True
31,C1398,31.0,Galle,82745.0,0.68,7.0,1,68.0,True
502,C1359,53.0,Kandy,106577.0,0.68,4.0,1,68.0,True
508,C1214,52.0,Kandy,81812.0,0.66,2.0,0,66.0,True
186,C1157,44.0,Jaffna,70139.0,0.65,6.0,0,65.0,False
167,C1016,42.0,Colombo,92111.0,0.62,1.0,0,62.0,True
484,C1252,26.0,Negombo,79909.0,0.62,11.0,0,62.0,True
117,C1361,41.0,Galle,71673.0,0.59,2.0,1,59.0,False
113,C1255,24.0,Galle,52085.0,0.59,2.0,0,59.0,False
420,C1303,41.0,Jaffna,88555.0,0.59,1.0,0,59.0,True


In [78]:
filtered=df[(df["city"]=="Colombo") & (df["score"]>=0.7) & (df["purchased"]==1)]
print(filtered)

Empty DataFrame
Columns: [customer_id, age, city, income, score, purchases, purchased, score_pct, high_value]
Index: []


In [81]:
print(filtered.shape[0])

0


In [83]:
df.describe(include='all')

,customer_id,age,city,income,score,purchases,purchased,score_pct,high_value
count,490,490.000000,490,490.000000,490.000000,490.000000,490.000000,490.000000,490
unique,490,NaN,5,NaN,NaN,NaN,NaN,NaN,2
top,C1217,NaN,Colombo,NaN,NaN,NaN,NaN,NaN,False
freq,1,NaN,213,NaN,NaN,NaN,NaN,NaN,255
mean,NaN,41.306122,NaN,74338.255102,0.024193,5.469388,0.308163,2.419294,NaN
std,NaN,12.822214,NaN,21395.558564,0.104079,3.481779,0.462206,10.407860,NaN
min,NaN,18.000000,NaN,18000.000000,0.000880,0.000000,0.000000,0.088000,NaN
25%,NaN,31.000000,NaN,61311.250000,0.005070,2.000000,0.000000,0.507000,NaN
50%,NaN,41.000000,NaN,75252.000000,0.006250,6.000000,0.000000,0.625000,NaN
75%,NaN,52.750000,NaN,87185.250000,0.007387,8.000000,1.000000,0.738750,NaN


In [107]:
df["city"].value_counts()


city
Colombo    213
Kandy      101
Galle       86
Jaffna      57
Negombo     33
Name: count, dtype: int64

In [109]:
df["purchased"].value_counts()

purchased_balance = df['purchased'].value_counts(normalize=True) * 100
print("\nPurchased class balance (%):")
print(purchased_balance)


Purchased class balance (%):
purchased
0    69.183673
1    30.816327
Name: proportion, dtype: float64


In [96]:
print(df.groupby("city")["score"].mean())

city
Colombo    0.011346
Galle      0.038862
Jaffna     0.027590
Kandy      0.036785
Negombo    0.024479
Name: score, dtype: float64


In [110]:
print(df.groupby("city")[["income","purchases"]].mean())

               income  purchases
city                            
Colombo  75135.070423   5.488263
Galle    76360.232558   5.348837
Jaffna   72495.035088   5.456140
Kandy    71614.227723   5.376238
Negombo  75446.696970   5.969697


In [111]:
print(
    df.groupby("city").agg(
        row_count = ("city","count"),
        mean_age = ("age","mean"),
        mean_score = ("score","mean"),
        purchase_rate = ("purchased","mean")
    )
)

         row_count   mean_age  mean_score  purchase_rate
city                                                    
Colombo        213  41.647887    0.011346       0.314554
Galle           86  39.383721    0.038862       0.325581
Jaffna          57  40.912281    0.027590       0.263158
Kandy          101  41.415842    0.036785       0.277228
Negombo         33  44.454545    0.024479       0.393939


### 4 Bullet-Point Findings for a Future ML Model

1. **Strong Predictor (`score`):** The `score` variable has a strong positive correlation ($r \approx 0.54$) with the purchase decision. Purchased customers have a much higher average score ($\approx 0.75$) than non-purchased customers ($\approx 0.53$).
2. **Class Imbalance:** The target `purchased` is imbalanced (approx. 31% purchase rate vs 69% non-purchase rate). The future model must optimize for metrics like F1-Score or ROC-AUC instead of simple accuracy.
3. **Non-Linear Demographics:** Key features like `age` ($r \approx 0.01$) and `income` ($r \approx 0.07$) show very low linear correlations. Tree-based models (e.g., XGBoost) or non-linear feature engineering are needed to capture their interactions.
4. **Geographic Variation:** Standarized city categories show distinct purchase behavior. Negombo has the highest purchase rate ($\approx 39.4\%$), whereas Jaffna has the lowest ($\approx 26.3\%$). Categorical encoding will capture these regional trends.